### ISSUE TITLE
Exceptions + LTO lead to "wasm-ld: undefined symbol: __threwValue"

### ISSUE THREAD

**Issue by kripken** [emscripten-core/emscripten#10066]
We don't have a small testcase for this, but it seems to be something with LTO affecting what exceptions code remains. Perhaps the exceptions-lowering pass in the LLVM backend needs to be run again after link, or something like that. Can maybe investigate this by seeing where that symbol is added, and how.

However, I assume this will go away with native wasm exceptions anyhow. cc @aheejin

**Comment by sbc100** [comment]
I believe this kind of thing can happen when LTO causes new undefined symbol to enter the build.  It not supposed to be able to do that.

IIRC this is why compiler-rt should never partake in LTO since it contains the symbols that compiler is known to generate.   I'll take a look..

**Comment by GregPerez83** [comment]
Hi @kripken and @sbc100.  I don't have a minimal repro, but I do have some additional information.  For my project, this occurs when I link to a static library that was compiled with "-flto" and the main executable is linked with "-s WASM_OBJECT_FILES=0 --llvm-lto 1".  When I remove "-flto" when compiling the lib (OpenCV), the issue goes away.

**Comment by sbc100** [comment]
Does adding `-u __threwValue` to the link line fix the problem?

**Comment by Akaricchi** [comment]
@sbc100 I seem to be having an identical problem with Taisei. I've tried `-u __threwValue` just now, but looks like it's not recognized:

```
[614/615] /data/webshit/emroot/emcc /data/webshit/ports/taisei/src/taisei.bc -s WASM_OBJECT_FILES=0 --llvm-lto 1 -flto --pre-js ../../../git/taisei/emscripten/preamble.js --pre-js resources/bundle_gfx.js --pre-js resources/bundle_misc.js --pre-js resources/bundle_bgm.js --pre-js resources/bundle_sfx.js --shell-file ../../../git/taisei/emscripten/shell.html -Wno-implicit-fallthrough -Wno-ignored-optimization-argument -v -O3 -s ALLOW_MEMORY_GROWTH=1 -s ENVIRONMENT=web -s EXIT_RUNTIME=0 -s 'EXPORTED_FUNCTIONS=["_main", "_vfs_sync_callback", "_SDL_LoadWAV_RW"]' -s 'EXTRA_EXPORTED_RUNTIME_METHODS=["ccall"]' -s FILESYSTEM=1 -s FORCE_FILESYSTEM=1 -s GL_POOL_TEMP_BUFFERS=0 -s GL_PREINITIALIZED_CONTEXT=1 -s LZ4=1 -s MAX_WEBGL_VERSION=2 -s MIN_WEBGL_VERSION=2 -s MODULARIZE=0 -s STRICT=1 -s TOTAL_MEMORY=268435456 -s WASM=1 -lGL -legl.js -lidbfs.js -s ASYNCIFY -O3 -g0 -s ASSERTIONS=0 -s GL_TRACK_ERRORS=0 --closure 1 -u __threwValue -o src/taisei.html
shared:INFO: (Emscripten: Running sanity checks)
shared:ERROR: __threwValue: No such file or directory ("__threwValue" was expected to be an input file, based on the commandline arguments provided)
```

Am I doing it wrong?

**Comment by sbc100** [comment]
Sorry `-u __threwValue` is the linker flag.   There are a few ways you could add them:

1.  Pass them to emcc witgh `-Wl,-u,__threwValue`.  If you do this you will also need to add those flags to SUPPORTED_LLD_LINKER_FLAGS in emcc.py.

2. Run the linker command yourself and add `-u __threwValue` directly.  To get a linker command that is repeatable you may need to run with EMCC_DEBUG=1 which keeps temporary files around.

3. Hardcode the flags in `tools/shared.py:link_lld`

**Comment by sbc100** [comment]
If either of you can great a minimal repro case that would be awesome.

**Comment by Akaricchi** [comment]
Just tried passing it directly to the linker, it seems to have worked (as in, it links). I haven't tested whether the result is actually functional yet though.

**Comment by Akaricchi** [comment]
Ok, just had a test run. The game seems to be working perfectly. Note however that I am not using C++ exceptions at all (I assume `__threwValue` is part of that). The entire code base is in C; there is some minor setjmp/longjmp usage for error handling, which my run may not have triggered. I'll try doing some more thorough testing in the next few days, and maybe come up with a reduced test case.

**Comment by sbc100** [comment]
Yes __threwValue is a symbol that llvm can generate for both setjmp/longjmp and exceptions (since exceptions today is based on the same tricks).

**Comment by sbc100** [comment]
I think I have a fix, I just need to write a test case to prove it.

**Comment by sbc100** [comment]
I'm now having trouble producing this locally.  Could you send the full `wasm-ld` command line you are seeing.

Even better if you could attache the full output running the linker with `-mllvm -debug-only=lld`?

**Comment by Akaricchi** [comment]
hmm

```
$ "/opt/llvm/bin/wasm-ld" -o /tmp/emscripten_temp/taisei.wasm --allow-undefined --lto-O3 /data/webshit/ports/taisei/src/taisei.bc -L/data/git/emscripten/system/lib -L/data/webshit/emsdk/.emscripten_cache/wasm-bc /data/webshit/emsdk/.emscripten_cache/wasm-bc/libc.bc /data/webshit/emsdk/.emscripten_cache/wasm-bc/libcompiler_rt.bc /data/webshit/emsdk/.emscripten_cache/wasm-bc/libc-wasm.bc /data/webshit/emsdk/.emscripten_cache/wasm-bc/libgl-webgl2.bc /data/webshit/emsdk/.emscripten_cache/wasm-bc/libpthread_stub.bc /data/webshit/emsdk/.emscripten_cache/wasm-bc/libsockets.bc /data/webshit/emsdk/.emscripten_cache/wasm-bc/libdlmalloc.a /data/webshit/emsdk/.emscripten_cache/wasm-bc/libc_rt_wasm.a --import-memory --import-table -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --export __wasm_call_ctors --export __data_end --export main --export vfs_sync_callback --export SDL_LoadWAV_RW --export malloc --export free --export setThrew --export __errno_location --export emscripten_builtin_memalign --export memalign --export _get_tzname --export _get_daylight --export _get_timezone --export emscripten_GetProcAddress --export emscripten_builtin_free --export _get_environ --export realloc --export strstr -z stack-size=5242880 --initial-memory=268435456 --no-entry --global-base=1024 -mllvm -debug-only=lld                                                                            
 -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --export __wasm_call_ctors --export __data_end --export main --export vfs_sync_callback --export SDL_LoadWAV_RW --export malloc --export free --export setThrew --export __errno_location --export emscripten_builtin_memalign --export memalign --export _get_tzname --export _get_daylight --export _get_timezone --export emscripten_GetProcAddress --export emscripten_builtin_free --export _get_environ --export realloc --export strstr -z stack-size=5242880 --initial-memory=268435456 --no-entry --global-base=1024 -mllvm -debug-only=lld

wasm-ld (LLVM option parsing): Unknown command line argument '-debug-only=lld'.  Try: 'wasm-ld (LLVM option parsing) --help'
wasm-ld (LLVM option parsing): Did you mean '--debug-pass=lld'?
```

**Comment by sbc100** [comment]
Thanks, I've manged to repro this now.   BTW, it looks like it only occurs with -O2 on the command line, is that true for you too?

**Comment by jpharvey** [comment]
Yes for me when i saw this it was -O2 and higher. I think it may also have been -Os but i cant remember its a while since i looked at it.

**Comment by sbc100** [comment]
Should be fixed in #10274
